### PR TITLE
Feature/30 integrate qoi metrics and feature/35-convergence-plots-to-work-with-QoI-Metrics

### PR DESCRIPTION
**Pull request by henrikstoklandberg**
Issue 30 integrate qoi metrics and 35 plots to work with QoI metrics added for crest_heights_north_sea doe and basic_example.

### CODE REVIEW

File: tutorials/basic_example.py [hunk 1]
```diff
@@ -656,24 +644,12 @@ def sobol_generator_run(_: Experiment) -> GeneratorRun:
 
 
 def look_ahead_generator_run(experiment: Experiment) -> GeneratorRun:
-    # Fist building model to get the transforms
```
Comment (swinter1) [inline tutorials/basic_example.py:659]: I don't think we can delete this yet. We nee to work out the transforms that will be built when the model_bridge is built for this dataset, so we can put them on the qoi_estimator. The qoi_estimator needs to have the transform corresponding to the model eing used with it.

I really hate who we have to do this transform stuff at the minute. If you have suggestion on a better way to do it I would be stocked to hear it.

Comment (swinter1) [reply tutorials/basic_example.py:659]: This will also need to be fixed in `doe.py`

File: tutorials/basic_example.py [hunk 2]
```diff
@@ -744,29 +737,33 @@ def plot_raw_ut_estimates(
     Returns:
         Axes: the ax with the plot.
     """
+    metrics = experiment.fetch_data()
+    qoi_metrics = metrics.df[metrics.df["metric_name"] == "QoIMetric"]
+
+    qoi_means = qoi_metrics["mean"]
+    qoi_sems = qoi_metrics["sem"]
+    var = qoi_sems**2
```
Comment (swinter1) [inline tutorials/basic_example.py:745]: we turn this into var and then back into std in `ax.fill_between`. do we need var?

**Review by swinter1 — CHANGES_REQUESTED** [review]
looking good! A few changes, but i like how its looking with the tracking metric

File: examples/crest_heights_north_sea/problem/doe.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+                _ = trial.mark_completed()
+
+        else:
+            generator_run = doe_generator(experiment)
+            trial = experiment.new_trial(generator_run=generator_run)
+            _ = trial.run()
+            _ = trial.mark_completed()
+        print(f"iter {i} done")
+
+    return
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:98]: done

File: examples/crest_heights_north_sea/problem/doe.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+from axtreme.utils import transforms
+
+torch.set_default_dtype(torch.float64)
+device = "cpu"
+
+# pyright: reportUnnecessaryTypeIgnoreComment=false
+
+# %%
+search_space = SEARCH_SPACE
+n_env_samples_per_period = N_ENV_SAMPLES_PER_PERIOD
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:40]: done

File: examples/crest_heights_north_sea/problem/doe.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+# can be in whatever format is supported by the QoI.
+n_env_samples = 1_000
+dataset = MinimalDataset(env_data)
+sampler = FixedRandomSampler(dataset, num_samples=n_env_samples, seed=10, replacement=True)
+dataloader = DataLoader(dataset, sampler=sampler, batch_size=256)
+
+posterior_sampler = UTSampler()
+# posterior_sampler = NormalIndependentSampler(torch.Size([n_posterior_samples])
+
+qoi_estimator = MarginalCDFExtrapolation(
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:114]: Done

File: examples/crest_heights_north_sea/problem/doe.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+# %%
+# Plot the surface of some trials
+fig_trial_warm_up = plot_gp_fits_2d_surface_from_experiment(exp_sobol, warm_up_runs)
+fig_trial_warm_up.show()
+fig_last_trial = plot_gp_fits_2d_surface_from_experiment(exp_sobol, n_iter)
+fig_last_trial.show()
+
+
+# %%
+# Define the model to use.
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:186]: done

File: tutorials/basic_example.py [hunk 7]
```diff
@@ -702,29 +678,47 @@ def look_ahead_generator_run(experiment: Experiment) -> GeneratorRun:
 # %% [markdown]
 # Run the DOE
 
+# %%
+# Create QoI tracking metric for tracking of the QoI estimate over the course of the experiment.
+qoi_metric = QoIMetric(
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply tutorials/basic_example.py:683]: done

File: tutorials/basic_example.ipynb
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply tutorials/basic_example.ipynb:1]: done

File: examples/crest_heights_north_sea/problem/doe.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+fig_last_trial = plot_gp_fits_2d_surface_from_experiment(exp_look_ahead, n_iter)
+fig_last_trial.show()
+
+
+# %% [markdown]
+# ### Plot the results
+
+
+# %%
+def plot_raw_ut_estimates(
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:330]: Done, had a quick look at ax.plot.trace.plot_objective_value_vs_trial_index, but I was not able to plot what I wanted really.

File: tutorials/basic_example.py [hunk 9]
```diff
@@ -702,29 +678,47 @@ def look_ahead_generator_run(experiment: Experiment) -> GeneratorRun:
...
+
 # %% [markdown]
 # ### Plot the results
 
 
 # %%
 def plot_raw_ut_estimates(
-    mean: torch.Tensor,
-    var: torch.Tensor,
+    experiment: Experiment,
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply tutorials/basic_example.py:721]: done

File: tutorials/basic_example.py (hunk 1 above)
Comment (henrikstoklandberg) [reply tutorials/basic_example.py:659]: done, no better way implemented

File: examples/crest_heights_north_sea/problem/doe.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (swinter1) [reply examples/crest_heights_north_sea/problem/doe.py:135]: please make `qoi_estimator` uppercase too.

File: examples/crest_heights_north_sea/problem/doe.py [hunk 10]
```diff
@@ -36,8 +41,26 @@
 # pyright: reportUnnecessaryTypeIgnoreComment=false
 
 # %%
-search_space = SEARCH_SPACE
-n_env_samples_per_period = N_ENV_SAMPLES_PER_PERIOD
+# TODO(@henrikstoklandberg): Update the search space to match the problem once we decide on the search space.
```
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/doe.py:44]: In future handy if you add a date to the comment (i know you added on at the end. 
```suggestion
# TODO(@henrikstoklandberg 2025-04-28): Update the search space to match the problem once we decide on the search space.
```

File: tutorials/basic_example.py [hunk 11]
```diff
@@ -660,7 +648,8 @@ def look_ahead_generator_run(experiment: Experiment) -> GeneratorRun:
     # TODO (se -2024-11-20): This refits hyperparameter each time, we don't want to do this.
     model_bridge_only_model = Models.BOTORCH_MODULAR(
         experiment=experiment,
-        data=experiment.fetch_data(),
+        data=experiment.fetch_data(metrics=list(experiment.optimization_config.metrics.values())),  # type: ignore  # noqa: PGH003
```
Comment (swinter1) [inline tutorials/basic_example.py:651]: Same question a as other comment

**Review by swinter1 — COMMENTED** [review]
A few final minor changes and then we are good to go

File: examples/crest_heights_north_sea/problem/doe.py (hunk 10 above)
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:44]: done

File: examples/crest_heights_north_sea/problem/doe.py [hunk 12]
```diff
@@ -36,8 +41,26 @@
...
+# For now a square search space is used, as we get Nan values from the simulator if we use the current search space in
+# problem.py as of 2025-05-28.
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:46]: fixed

File: examples/crest_heights_north_sea/problem/doe.py [hunk 13]
```diff
@@ -301,7 +289,8 @@ def look_ahead_generator_run(experiment: Experiment) -> GeneratorRun:  # noqa: D
     # TODO (se -2024-11-20): This refits hyperparameter each time, we don't want to do this.
     model_bridge_only_model = Models.BOTORCH_MODULAR(
         experiment=experiment,
-        data=experiment.fetch_data(),
+        data=experiment.fetch_data(metrics=list(experiment.optimization_config.metrics.values())),  # type: ignore  # noqa: PGH003
```
[the remark this answers is not in the document]
Comment (henrikstoklandberg) [reply examples/crest_heights_north_sea/problem/doe.py:292]: I have implemented the workaround now and made a ticket to fix this at a later stage.

File: tutorials/basic_example.py (hunk 11 above)
Comment (henrikstoklandberg) [reply tutorials/basic_example.py:651]: I have implemented the workaround now and made a ticket to fix this at a later stage.

**Review by swinter1 — APPROVED** [review]
